man/halt.8
'\" -*- coding: UTF-8 -*-
.\" Copyright (C) 1998-2001 Miquel van Smoorenburg.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
.\"
.\"{{{}}}
.\"{{{  Title
.TH HALT 8 "Nov 6, 2001" "sysvinit " "Linux System Administrator's Manual"
.\"}}}
.\"{{{  Name
.SH NAME
halt, reboot, poweroff \- stop the system
.\"}}}
.\"{{{  Synopsis
.SH SYNOPSIS
.B /sbin/halt
.RB [ \-d ]
.RB [ \-f ]
.RB [ \-h ]
.RB [ \-i ]
.RB [ \-n ]
.RB [ \-p ]
.RB [ \-w ]
.br
.B /sbin/reboot
.RB [ \-d ]
.RB [ \-f ]
.RB [ \-i ]
.RB [ \-k ]
.RB [ \-m\ <message> ]
.RB [ \-n ]
.RB [ \-w ]
.br
.B /sbin/poweroff
.RB [ \-d ]
.RB [ \-f ]
.RB [ \-h ]
.RB [ \-i ]
.RB [ \-n ]
.RB [ \-w ]
.\"}}}
.\"{{{  Description
.SH DESCRIPTION
\fBhalt\fP notes that the system is being brought down in the file
\fI/var/log/wtmp\fP, and then either tells the kernel to halt, reboot or
power-off the system.
.PP
If \fBhalt\fP or \fBreboot\fP is called when the system is
\fInot\fP in runlevel \fB0\fP or \fB6\fP, in other words when it's running
normally, \fBshutdown\fP will be invoked instead (with the \fB\-h\fP
or \fB\-r\fP flag).
For more info see the \fBshutdown\fP(8) manpage.
.PP
The rest of this manpage describes the behavior in runlevels 0
and 6, that is when the systems shutdown scripts are being run.
.\"}}}
.\"{{{  Options
.SH OPTIONS
.IP \fB\-d\fP
Don't write the wtmp record.
.IP \fB\-f\fP
Force halt or reboot, don't call \fBshutdown\fP(8).
.IP \fB\-h\fP
Put all hard drives on the system in stand-by mode
just before halt or power-off.
.IP \fB\-i\fP
Shut down all network interfaces just before \fBhalt\fR or \fBreboot\fR.
Warning: This may not work on interfaces which do not have an IP address
and should ideally be handled by a network manager service.
.IP \fB\-k\fP
Try to reboot using \fBkexec\fP, if kernel supports it.
.IP \fB\-m\fP
When running on a Linux system, any quoted parameter after the
\-m flag will be passed to the firmware to be interpreted at boot time.
Some systems, like the Raspberry Pi,
can use this to switch target partitions.
Please note, this parameter enables \-f,
forcing the reboot command to handle shutting down itself.
.IP \fB\-n\fP
Don't sync before reboot or halt.
Note that the kernel and storage drivers may still sync.
This implies \fB\-d\fP.
.IP \fB\-p\fP
When halting the system, switch off the power.
This is the default when \fBhalt\fR is called as \fBpoweroff\fP.
.IP \fB\-w\fP
Don't actually reboot or halt but only write the wtmp record
(in the \fI/var/log/wtmp\fP file).
.\"}}}
.\"{{{  Diagnostics
.SH DIAGNOSTICS
If you're not the superuser, you will get the message `must be superuser'.
.\"}}}
.\"{{{  Notes
.SH NOTES
Under older \fBsysvinit\fP releases, \fBreboot\fP and \fBhalt\fP should
never be called directly.
From release 2.74 on,
\fBhalt\fP and \fBreboot\fP invoke \fBshutdown\fP(8)
if the system is not in runlevel 0 or 6.
This means that if \fBhalt\fP or \fBreboot\fP cannot find out the current
runlevel
(for example, when \fI/var/run/utmp\fP hasn't been initialized correctly and
\fI/var/run/runlevel\fR does not exist)
\fBshutdown\fP will be called,
which might not be what you want.
Use the \fB\-f\fP flag if you want to do a hard \fBhalt\fP or \fBreboot\fP.
.PP
The \fB\-h\fP flag puts all hard disks in standby mode just before halt
or power-off.
Right now this is only implemented for IDE drives.
A side effect of putting the drive in stand-by mode is
that the write cache on the disk is flushed.
This is important for IDE drives,
since the kernel doesn't flush the write cache itself before power-off.
.PP
The \fBhalt\fP program uses \fI/proc/ide/hd*\fR to find all IDE disk devices,
which means that \fI/proc\fP needs to be mounted when \fBhalt\fP or
\fBpoweroff\fP is called or the \fB\-h\fP switch will do nothing.
.\"}}}
.\"{{{  Author
.SH AUTHOR
.MT miquels@\:cistron\:.nl
Miquel van Smoorenburg
.ME
.\"}}}
.\"{{{  See also
.SH "SEE ALSO"
.BR shutdown (8),
.BR init (8)
.\"}}}


doc/Changelog
sysvinit (3.11) unreleased; urgency=low

    * Some escape characters were included in the inittab manual page, but not displayed
    by the "man" command because they were not (ironically) properly escaped. This has been
    fixed.

    * Enabled chaining commands together in the inittab file. This allows the admin
      to run commands like "task1 && task2" or "task2 || task2" from the inittab file.

    * Fix typoes in halt manual page. Fixes provided by Bjarni Ingi Gislason.


sysvinit (3.10) released; urgency=low

    *  When the user executes "machinectl stop", systemd sends SIGRTMIN+4 to PID 1 
       in the container, and expects that to initiate a graceful shutdown (power-off).
       SysV init now catches this signal and initiates a shutdown (shutdown -hP now).
       - floppym provided patch to accomplish this.

    * Fix issue in bootlogd which could cause the service to enter an endless loop
      (and use too much CPU) when it is able to open a device for writing, but not actually
      able to write to it. This resulted in bootlogd closing and re-opening the device over
      and over. Now bootlogd should simply fail gracefully when it cannot write to an open
      file/device.

    * Fix formatting in shutdown.8 manual page. Cleaned up whitespace and special characters.


sysvinit (3.09) released; urgency=low

    * Patch for man/Makefile to fix the clean recipe.
      Provided by Lucas Nussabaum and Mark Hindley

    * On Linux systems, allow reboot command to pass a message
      to the system firmware during the restart. This is
      accomplished with the -m flag.

    * Patch from kraj which allows hddown to compile
      when musl is the C library.

sysvinit (3.08) released; urgency=low
    * floppym provided patch which causes the halt command
      to call "shutdown -h -H" instead of "shutdown -h" when
      halt is invoked without parameters. This forces the shutdown
      command to set the INIT_HALT variable and assume, unless other
      conditions apply, that the "halt" call really wants to halt the
      machine and INIT_HALT should be set. In other words we
      assume halt wants to halt unless told otherwise.
      Addresses downstream Gentoo bug ID 911257.

    * Applied a patch from floppm which adds kexec option to the halt command.

    * Updated halt documentation and help output to display parameters in
      alphabetical order.


sysvinit (3.07) released; urgency=low
    * Fixed killall5 so that processes in the omit list are
      not sent any signals, including SIGSTOP.
    * Fixed usage message for killall5 to be more accurate.
    * pidof was not returning PIDs of programs which were launched
      using a symbolic link. (ie /tmp/sleep when /tmp/sleep links to /usr/bin/sleep).
      This is now fixed as we check both the realpath and symbolic path for processes.
      In other words, "pidof /tmp/sleep" and "pidof /usr/bin/sleep" will return
      the same PIDs when /tmp/sleep is a symbolic link to /usr/bin/sleep.
    * Fixed memory initialization error in pidof. Fix provided by Markus Fischer.
    * Accepted patch from Mark Hindley  which avoids clearing realpath information
      in pidof when trying to find matching executables.


sysvinit (3.06) released; urgency=low
    * Mark Hindley fixed typo in es.po
    * Mark Hindley cleaned up translation code in src/Makefile. 
    * Drop sulogin from Debian build. Removed libcrypt-dev dependency.
    * Fixed pt translation pages which were failing due to mis-matched
      open/close tags.
    * Makefile now respects ROOT prefix when setting up pidof-to-killall5
      symbolic link.
    * Removed redundant translation files from man directory.
    * Makefile now respects DESTDIR. User can specify either ROOT= or DESTDIR=
      to set install prefix.


sysvinit (3.05) released; urgency=low
    * Helge Kreutzmann provided updated Makefile for
      translation of manual pages. This has been added
      to the man directory.
    * Added sys/sysmacros.h include in mountpoint.c to fix
      compiler errors on systems where major/minor macros were not defined.
    * Applied patches from Mark Hindley to clean up man page
      Makefile, translations and installs of new man pages.
    * Remove reliance on linux/fs.h as it conflicts with glibc 2.36.
      Patch provided by lucascars.

sysvinit (3.04) released; urgency=low
    * Mark Hindley supplied patch to make bootlogd compile
      on GNU Hurd systems. Was missing major/minor macro.

sysvinit (3.03) released; urgency=low
    * Fixed formatting in init.8 man page.
      Patch provided by Mark Hindley.
    * Fixed compile error on GNU Hurd and musl-base systems. Due to missing MAX_PATH definition.
      Error reported by Mark Hindly.

sysvinit (3.02) released; urgency=low
    * Added q and Q flags to synopsis in shutdown manual page.
    * Applied fixes for markup and spacing in manual pages.
      Patch provided by Mario Blattermann.
    * Added translation framework (po4a) from Mario Blttermann.
    * Added Makefile for man/ directory. Will handle translations
      and substitutions.
    * Applied new translations for multiple languages from Mario Blattermann.
    * Added ability to use "@" symbol in command named in the inittab file. This
      treats commands as literal and does not launch a shell to interpret them.
    * Updated inittab manual page to include overview of symbols which trigger
      a shell interpretor and how to disable them using the @ symbol.
    * Introduced change which adds error checking in bootlogd when performing
      chdir(). - Provided by Alexander Vickberg 
    * Add check for console using TIOCGDEV on Linux systems in bootlogd to
      make finding console more robust. - Provided by Alexander Vickberg
 

sysvinit (3.01) released; urgency=low
    * Default to showing processes in the uninterruptable state (D).
    The -z flag no longer affects whether processes in D state are shown.

src/halt.c
/*
 * Halt		Stop the system running.
 *		It re-enables CTRL-ALT-DEL, so that a hard reboot can
 *		be done. If called as reboot, it will reboot the system.
 *
 *		If the system is not in runlevel 0 or 6, halt will just
 *		execute a "shutdown -h" to halt the system, and reboot will
 *		execute an "shutdown -r". This is for compatibility with
 *		sysvinit 2.4.
 *
 * Usage:	halt [-n] [-w] [-d] [-f] [-h] [-i] [-p] [-k]
 *		-n: don't sync before halting the system
 *		-w: only write a wtmp reboot record and exit.
 *		-d: don't write a wtmp record.
 *		-f: force halt/reboot, don't call shutdown.
 *		-h: put harddisks in standby mode
 *		-i: shut down all network interfaces.
 *		-p: power down the system (if possible, otherwise halt).
 *		-k: reboot the system using kexec.
 *              -m: allow user to pass in message to Linux kernel to enable switching partitions
 *
 *		Reboot and halt are both this program. Reboot
 *		is just a link to halt. Invoking the program
 *		as poweroff implies the -p option.
 *
 * Author:	Miquel van Smoorenburg, miquels@cistron.nl
 *
 * Version:	2.86,  30-Jul-2004
 *
 *		This file is part of the sysvinit suite,
 *		Copyright (C) 1991-2004 Miquel van Smoorenburg.
 *
 *		This program is free software; you can redistribute it and/or modify
 *		it under the terms of the GNU General Public License as published by
 *		the Free Software Foundation; either version 2 of the License, or
 *		(at your option) any later version.
 *
 *		This program is distributed in the hope that it will be useful,
 *		but WITHOUT ANY WARRANTY; without even the implied warranty of
 *		MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *		GNU General Public License for more details.
 *
 *		You should have received a copy of the GNU General Public License
 *		along with this program; if not, write to the Free Software
 *		Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <stdlib.h>
#include <utmp.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <sys/times.h>
#include <time.h>
#include <signal.h>
#include <stdio.h>
#include <getopt.h>
#ifdef __linux__
#include <linux/reboot.h>
#include <sys/syscall.h>
#endif
#include "reboot.h"
#include "runlevellog.h"

char *Version = "@(#)halt  2.86  31-Jul-2004 miquels@cistron.nl";
char *progname;

#define KERNEL_MONITOR	1 /* If halt() puts you into the kernel monitor. */
#define RUNLVL_PICKY	0 /* Be picky about the runlevel */

extern int ifdown(void);
extern int hddown(void);
extern int hdflush(void);
extern void write_wtmp(char *user, char *id, int pid, int type, char *line);

/*
 *	Send usage message.
 */
void usage(void)
{
	fprintf(stderr, "usage: %s [-d] [-f] [-h] [-i] [-n] [-w]%s%s\n",
		progname,
		strcmp(progname, "halt") ? "" : " [-p]",
		strcmp(progname, "reboot") ? "" : " [-k] [-m <message>]");
	fprintf(stderr, "\t-d: don't write a wtmp record.\n");
	fprintf(stderr, "\t-f: force halt/reboot, don't call shutdown.\n");
	fprintf(stderr, "\t-h: put harddisks in standby mode.\n");
	fprintf(stderr, "\t-i: shut down all network interfaces.\n");
	fprintf(stderr, "\t-m: pass message from user to reboot process.\n");
	fprintf(stderr, "\t-n: don't sync before halting the system\n");
	fprintf(stderr, "\t-w: only write a wtmp reboot record and exit.\n");
	if (!strcmp(progname, "halt"))
		fprintf(stderr, "\t-p: power down the system (if possible, otherwise halt).\n");
	if (!strcmp(progname, "reboot"))
		fprintf(stderr, "\t-k: reboot the system using kexec.\n");
	exit(1);
}

/*
 *	See if we were started directly from init.
 *	Get the runlevel from /var/run/utmp or the environment.
 *      Or the /var/run/runlevel log.
 */
int get_runlevel(void)
{
	struct utmp *ut;
	char *r;
        int runlevel, status;

#if RUNLVL_PICKY
	time_t boottime;
#endif

	/*
	 *	First see if we were started directly from init.
	 */
	if (getenv("INIT_VERSION") && (r = getenv("RUNLEVEL")) != NULL)
		return *r;

	/*
	 *	Hmm, failed - read runlevel from /var/run/utmp..
	 */
#if RUNLVL_PICKY
	/*
	 *	Get boottime from the kernel.
	 */
	time(&boottime);
	boottime -= (times(NULL) / HZ);
#endif

	/*
	 *	Find runlevel in utmp.
	 */
	setutent();
	while ((ut = getutent()) != NULL) {
#if RUNLVL_PICKY
		/*
		 *	Only accept value if it's from after boottime.
		 */
		if (ut->ut_type == RUN_LVL && ut->ut_time > boottime)
			return (ut->ut_pid & 255);
#else
		if (ut->ut_type == RUN_LVL)
			return (ut->ut_pid & 255);
#endif
	}
	endutent();

        /* Did not find utmp entry, try to read from log file */
        status = Read_Runlevel_Log(&runlevel);
        if (status)
           return runlevel;

	/* This should not happen but warn the user! */
	fprintf(stderr, "WARNING: could not determine runlevel"
		" - doing soft %s\n", progname);
	fprintf(stderr, "  (it's better to use shutdown instead of %s"
		" from the command line)\n", progname);

	return -1;
}

/*
 *	Switch to another runlevel.
 */
void do_shutdown(char *fl, int should_poweroff, char *tm)
{
	char *args[9];
	int i = 0;

	args[i++] = "shutdown";
	args[i++] = fl;
        if (! strcmp(fl, "-h"))
           args[i++] = (should_poweroff ? "-P" : "-H");
	if (tm) {
		args[i++] = "-t";
		args[i++] = tm;
	}
	args[i++] = "now";
	args[i++] = NULL;

	execv("/sbin/shutdown", args);
	execv("/etc/shutdown", args);
	execv("/bin/shutdown", args);

	perror("shutdown");
	exit(1);
}

/*
 *	Main program.
 *	Write a wtmp entry and reboot cq. halt.
 */
int main(int argc, char **argv)
{
	int do_reboot = 0;
	int do_sync = 1;
	int do_wtmp = 1;
	int do_nothing = 0;
	int do_hard = 0;
	int do_ifdown = 0;
	int do_hddown = 0;
	int do_poweroff = 0;
	int do_kexec = 0;
        char *user_message = NULL;
	int c;
	char *tm = NULL;

	/*
	 *	Find out who we are
	 */
	/* Remove dash passed on in argv[0] when used as login shell. */
	if (argv[0][0] == '-') argv[0]++;
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	if (!strcmp(progname, "reboot")) do_reboot = 1;
	if (!strcmp(progname, "poweroff")) do_poweroff = 1;

	/*
	 *	Get flags
	 */
	while((c = getopt(argc, argv, ":ihdfm:npwkt:")) != EOF) {
		switch(c) {
			case 'n':
				do_sync = 0;
				do_wtmp = 0;
				break;
			case 'w':
				do_nothing = 1;
				break;
			case 'd':
				do_wtmp = 0;
				break;
			case 'f':
				do_hard = 1;
				break;
			case 'i':
				do_ifdown = 1;
				break;
			case 'h':
				do_hddown = 1;
				break;
			case 'p':
				do_poweroff = 1;
				break;
			case 'k':
				do_kexec = 1;
				break;
                        case 'm':
                                user_message = optarg;
                                do_hard = 1;
                                break;
			case 't':
				tm = optarg;
				break;
			default:
				usage();
		}
	 }
	if (argc != optind) usage();

	if (geteuid() != 0) {
		fprintf(stderr, "%s: must be superuser.\n", progname);
		exit(1);
	}

	if (chdir("/")) {
		fprintf(stderr, "%s: chdir(/): %m\n", progname);
		exit(1);
	}

	if (!do_hard && !do_nothing) {
		c = get_runlevel();

		/*
		 *	 We can't reboot using kexec through this path.
		 */
		if (c != '6' && do_reboot && do_kexec) {
			fprintf(stderr, "ERROR: using -k at this"
				" runlevel requires also -f\n"
				"  (You probably want instead to reboot"
				" normally and let your reboot\n"
				"   script, usually /etc/init.d/reboot,"
				" specify -k)\n");
			exit(1);
		}

		/*
		 *	See if we are in runlevel 0 or 6.
		 */
		if (c != '0' && c != '6')
			do_shutdown(do_reboot ? "-r" : "-h", do_poweroff, tm);
	}

	/*
	 *	Record the fact that we're going down
	 */
	if (do_wtmp)
		write_wtmp("shutdown", "~~", 0, RUN_LVL, "~~");

	/*
	 *	Exit if all we wanted to do was write a wtmp record.
	 */
	if (do_nothing && !do_hddown && !do_ifdown) exit(0);

	if (do_sync) {
		sync();
		/* Sync should be fine on its own for making sure data is written.
                   We probably call shutdown after this anyway to clean up.
                   -- Jesse
                sleep(2);
                */
	}

	if (do_ifdown)
		(void)ifdown();

	if (do_hddown)
		(void)hddown();
	else
		(void)hdflush();

	if (do_nothing) exit(0);

	if (do_reboot) {
		/*
		 *	kexec or reboot
		 */
		if (do_kexec)
			init_reboot(BMAGIC_KEXEC);

                /* Regular reboot, but with special user message */
                #ifdef __linux__
                if (user_message)
                {
                   syscall(SYS_reboot, LINUX_REBOOT_MAGIC1, LINUX_REBOOT_MAGIC2, 
                           LINUX_REBOOT_CMD_RESTART2, user_message);
                }
                #endif
		/*
		 *	Fall through if failed
		 */
		init_reboot(BMAGIC_REBOOT);
	} else {
		/*
		 *	Turn on hard reboot, CTRL-ALT-DEL will reboot now
		 */
#ifdef BMAGIC_HARD
		init_reboot(BMAGIC_HARD);
#endif

		/*
		 *	Stop init; it is insensitive to the signals sent
		 *	by the kernel.
		 */
		kill(1, SIGTSTP);

		/*
		 *	Halt or poweroff.
		 */
		if (do_poweroff)
			init_reboot(BMAGIC_POWEROFF);
		/*
		 *	Fallthrough if failed.
		 */
		init_reboot(BMAGIC_HALT);
	}

	/*
	 *	If we return, we (c)ontinued from the kernel monitor.
	 */
#ifdef BMAGIC_SOFT
	init_reboot(BMAGIC_SOFT);
#endif
	kill(1, SIGCONT);

	exit(0);
}
